### PR TITLE
Update readme and tests for ulog

### PR DESCRIPTION
**Pull request by anuptalwalkar**
Updated README documentation after zap.1.0.0 is merged.

### CODE REVIEW

File: ulog/README.md [hunk 1]
```diff
@@ -22,13 +22,12 @@ import "go.uber.org/fx/ulog"
 func main() {
   // Configure logger with configuration preferred by your service
   logConfig := ulog.Configuration{}
-  logBuilder := ulog.Builder().WithConfiguration(&logConfig)
 
-  // Build ulog.Log from logBuilder
-  log := lobBuilder.Build()
+  // Build logger from logConfig object
+  logger, err := svc.logConfig.Build(zap.Hooks(ulog.Metrics(svc.metrics)))
```
Comment (sectioneight) [inline ulog/README.md]: there's no `svc` in scope here...

File: ulog/README.md [hunk 2]
```diff
@@ -53,16 +52,20 @@ For example, the following piece of code:
...
+import (
+  "context"
 
+  "go.uber.org/fx/ulog"
+)
 func main() {
-  log := ulog.Logger()
-  log.Info("My info message")
-  log.Info("Info with context", "customer_id", 1234)
+
```
Comment (sectioneight) [inline ulog/README.md]: remove?

File: ulog/README.md [hunk 3]
```diff
@@ -22,13 +22,12 @@ import "go.uber.org/fx/ulog"
...
-  logBuilder := ulog.Builder().WithConfiguration(&logConfig)
 
-  // Build ulog.Log from logBuilder
-  log := lobBuilder.Build()
+  // Build logger from logConfig object
+  logger, err := svc.logConfig.Build(zap.Hooks(ulog.Metrics(svc.metrics)))
 
   // Use logger in your service
-  log.Info("Message describing logging reason", "key", "value")
+  log.Infow("Message describing logging reason", "key", "value")
```
Comment (sectioneight) [inline ulog/README.md]: should this be `logger` instead of `log` now?

File: ulog/README.md (hunk 1 above)
Comment (sectioneight) [inline ulog/README.md]: also `logConfig` isn't going to be exported, how are people going to access this? or maybe i just am mis-understanding what `svc`

Comment (anuptalwalkar) [inline ulog/README.md]: uh oh, you are right. Its the logConfig object, not related to svc. Fixing it.

File: ulog/README.md [hunk 4]
```diff
@@ -22,13 +22,12 @@ import "go.uber.org/fx/ulog"
 func main() {
   // Configure logger with configuration preferred by your service
   logConfig := ulog.Configuration{}
-  logBuilder := ulog.Builder().WithConfiguration(&logConfig)
 
-  // Build ulog.Log from logBuilder
-  log := lobBuilder.Build()
+  // Build logger from logConfig object
+  log, err := logConfig.Build(zap.Hooks(ulog.Metrics(svc.metrics)))
```
Comment (sectioneight) [inline ulog/README.md]: where does `svc` come and how are you accessing its `.metrics` (unexported) member?

Comment (anuptalwalkar) [inline ulog/README.md]: This should be part of service initialization, but not needed. I'll add simple build here.

File: ulog/log_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+func TestDefaultLogger(t *testing.T) {
+	cfg := DefaultConfiguration()
+	assert.NotNil(t, cfg)
+	assert.Equal(t, []string{"stdout"}, cfg.OutputPaths)
+}
+
+func TestSetLogger(t *testing.T) {
+
```
Comment (glibsm) [inline ulog/log_test.go]: mega blocking: extra empty line

Comment (anuptalwalkar) [inline ulog/log_test.go]: :(

File: ulog/README.md [hunk 6]
```diff
@@ -1,16 +1,15 @@
 # Logging package
 
 `package ulog` provides an API wrapper around the logging library
-[zap](https://github.com/uber-go/zap). `package ulog` uses the builder pattern
-to instantiate the logger. With `LogBuilder` you can perform pre-initialization
+[zap](https://github.com/uber-go/zap). `package ulog` uses zap's builder
+to instantiate the logger. With zap's `Builder` you can perform pre-initialization
```
Comment (akshayjshah) [inline ulog/README.md]: There is no `Builder` type, and there's no longer any support for injecting custom loggers.

File: ulog/README.md [hunk 7]
```diff
@@ -1,16 +1,15 @@
...
 setup by injecting configuration, custom logger, and log level prior to building
-the usable `ulog.Log` object.
+the usable `zapcore.Log` object.
```
Comment (akshayjshah) [inline ulog/README.md]: There's no `Log` type.

File: ulog/README.md [hunk 8]
```diff
@@ -53,16 +51,19 @@ For example, the following piece of code:
...
+import (
+  "context"
 
+  "go.uber.org/fx/ulog"
+)
 func main() {
-  log := ulog.Logger()
-  log.Info("My info message")
-  log.Info("Info with context", "customer_id", 1234)
+  log := ulog.Logger(context.Background())
```
Comment (akshayjshah) [inline ulog/README.md]: This shouldn't be a pattern we advertise - it's faster and functionally identical to just use `zap.L()`. We shouldn't encourage allocating a context just to access an already-available global.

Also, `ulog.Logger` returns a `*zap.Logger`. If you want a sugared logger, you have to use `ulog.Sugar`. Why not make these examples using the built-in testing package? Then you'll at least be guaranteed that they compile.

File: ulog/README.md [hunk 9]
```diff
@@ -1,16 +1,15 @@
 # Logging package
 
 `package ulog` provides an API wrapper around the logging library
```
Comment (glibsm) [inline ulog/README.md]: This is no longer true, right? We don't wrap anything

File: ulog/README.md [hunk 10]
```diff
@@ -22,13 +21,12 @@ import "go.uber.org/fx/ulog"
...
-  logBuilder := ulog.Builder().WithConfiguration(&logConfig)
 
-  // Build ulog.Log from logBuilder
-  log := lobBuilder.Build()
+  // Build logger from logConfig object
+  log, err := logConfig.Build()
 
   // Use logger in your service
-  log.Info("Message describing logging reason", "key", "value")
+  log.Infow("Message describing logging reason", "key", "value")
```
Comment (akshayjshah) [inline ulog/README.md]: This doesn't work: `log` is a `*zap.Logger`, not a `*zap.SugaredLogger`.

File: ulog/README.md (hunk 9 above)
Comment (anuptalwalkar) [inline ulog/README.md]: We still provide ulog.Logger(ctx) which is a major use case explained in context logging.

File: ulog/README.md [hunk 11]
```diff
@@ -1,16 +1,15 @@
...
+the usable `zapcore.Log` object.
 
-`ulog.Log` interface provides a few benefits:
+`ulog` provides a few benefits:
 
-- Decouple services from the logger used underneath the framework
-- Easy to use API for logging
-- Easily swappable backend logger without changing the service
+- Coupling with zap's Logger APIs.
+- Context based logging access via ulog.Logger(ctx)
```
Comment (akshayjshah) [inline ulog/README.md]: Suggest also pointing out that devs can use `ulog.Trace(ctx)` directly as a field.

File: ulog/log_test.go [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+func TestDefaultLogger(t *testing.T) {
+	cfg := DefaultConfiguration()
+	assert.NotNil(t, cfg)
+	assert.Equal(t, []string{"stdout"}, cfg.OutputPaths)
+}
+
+func TestSetLogger(t *testing.T) {
+	zaplogger := zapcore.NewNopCore()
+	SetLogger(zap.New(zaplogger))
```
Comment (akshayjshah) [inline ulog/log_test.go]: You *must* defer the returned undo function, or this test permanently alters the behavior of the standard library's `log` package.

Also, there's no need to explicitly construct a `zapcore.Core` - just use `zap.NewNop`.

File: ulog/README.md (hunk 7 above)
Comment (anuptalwalkar) [inline ulog/README.md]: good point. I was referring to the interface, but changed to `zap.Logger`.

File: ulog/README.md (hunk 8 above)
Comment (anuptalwalkar) [inline ulog/README.md]: The use of context is necessary to demonstrate context based logging (with tracer, etc available in context). Perhaps writing it in main is not wise. Will update it to some handler function that provides a context.

File: ulog/log_test.go (hunk 12 above)
Comment (anuptalwalkar) [inline ulog/log_test.go]: Wanted to keep zapcore to validate the object.

File: ulog/README.md (hunk 10 above)
Comment (anuptalwalkar) [inline ulog/README.md]: Changed the log to Sugar

File: ulog/README.md (hunk 9 above)
Comment (glibsm) [inline ulog/README.md]: Yeah, it's fine to provide that but we no longer wrap zap in any way though - that's old way

Comment (anuptalwalkar) [inline ulog/README.md]: fair point. I will update.

**Comment by glibsm** [comment]
what's up with the stuck coverage?

**Comment by anuptalwalkar** [comment]
does it depend on s3? kicking it off again.

File: ulog/README.md [hunk 13]
```diff
@@ -1,16 +1,21 @@
 # Logging package
 
-`package ulog` provides an API wrapper around the logging library
-[zap](https://github.com/uber-go/zap). `package ulog` uses the builder pattern
-to instantiate the logger. With `LogBuilder` you can perform pre-initialization
-setup by injecting configuration, custom logger, and log level prior to building
-the usable `ulog.Log` object.
+`package ulog` provides access to the logging library [zap](https://github.com/uber-go/zap)
+for the framework. `package ulog` uses zap's configuration wrapped
```
Comment (sectioneight) [inline ulog/README.md]: capitalize Zap?

File: ulog/README.md [hunk 14]
```diff
@@ -1,16 +1,21 @@
...
+configuration object, you can perform pre-initialization setup by injecting
+configuration, custom logger, and log level prior to building the usable `zap.Logger`.
 
-`ulog.Log` interface provides a few benefits:
+`ulog` provides a few benefits:
 
-- Decouple services from the logger used underneath the framework
-- Easy to use API for logging
-- Easily swappable backend logger without changing the service
+- Coupling with zap's Logger APIs.
```
Comment (glibsm) [inline ulog/README.md]: i don't get this point

File: ulog/log_test.go [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func TestDefaultLogger(t *testing.T) {
+	cfg := DefaultConfiguration()
```
Comment (alsamylkin) [inline ulog/log_test.go]: t.Parallel()?

File: ulog/README.md [hunk 16]
```diff
@@ -1,16 +1,21 @@
...
+- Context based logging access via `ulog.Logger(ctx)` for `zap.Logger` and `ulog.Sugar(ctx)` for Sugared logger
```
Comment (glibsm) [inline ulog/README.md]: "context aware"

File: ulog/log_test.go [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+	assert.NotNil(t, cfg)
+	assert.Equal(t, []string{"stdout"}, cfg.OutputPaths)
+}
+
+func TestSetLogger(t *testing.T) {
+	zaplogger := zapcore.NewNopCore()
+	undoFunc := SetLogger(zap.New(zaplogger))
+	defer undoFunc()
```
Comment (alsamylkin) [inline ulog/log_test.go]: Just defer SetLogger(zap.New(zaplogger))?

File: ulog/README.md (hunk 14 above)
Comment (anuptalwalkar) [inline ulog/README.md]: removed, since next two points cover this.

**Pull request merged**